File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/movie/query/MovieUpcomingQuery.java
package dev.reuss.tmdb.domain.movie.query;

import dev.reuss.tmdb.core.http.QueryParams;
import dev.reuss.tmdb.query.PagedQuery;
import dev.reuss.tmdb.query.QueryValidation;
import dev.reuss.tmdb.value.language.Language;
import dev.reuss.tmdb.value.region.Region;

/**
 * Query parameters for TMDB upcoming movie lists.
 *
 * <p>Language, page and region are optional. Page values are validated by
 * {@link QueryValidation#validatePage(Integer)}.</p>
 */
public final class MovieUpcomingQuery implements PagedQuery<MovieUpcomingQuery> {

    private Language language;
    private Integer page;
    private Region region;

    private MovieUpcomingQuery() {
    }

    public static MovieUpcomingQuery create() {
        return new MovieUpcomingQuery();
    }

    public MovieUpcomingQuery language(Language language) {
        this.language = language;
        return this;
    }

    @Override
    public MovieUpcomingQuery page(Integer page) {
        QueryValidation.validatePage(page);
        this.page = page;
        return this;
    }

    public MovieUpcomingQuery region(Region region) {
        this.region = region;
        return this;
    }

    @Override
    public QueryParams toQueryParams() {
        return QueryParams.create()
                .add("language", language)
                .add("page", page)
                .add("region", region);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/movie/query/MovieUpcomingQuery.kt
package dev.reuss.tmdb.domain.movie.query

import dev.reuss.tmdb.core.http.QueryParams
import dev.reuss.tmdb.query.PagedQuery
import dev.reuss.tmdb.query.QueryValidation
import dev.reuss.tmdb.value.language.Language
import dev.reuss.tmdb.value.region.Region

/**
 * Query parameters for TMDB upcoming movie lists.
 *
 * Language, page and region are optional. Page values are validated by
 * [QueryValidation.validatePage].
 */
class MovieUpcomingQuery private constructor() : PagedQuery<MovieUpcomingQuery> {
    private var language: Language? = null
    private var page: Int? = null
    private var region: Region? = null

    /** Sets the response language; `null` leaves the parameter unspecified. */
    fun language(value: Language?) =
        apply {
            language = value
        }

    /** Sets the one-based result page; `null` leaves the parameter unspecified. */
    override fun page(page: Int?) =
        apply {
            QueryValidation.validatePage(page)
            this.page = page
        }

    /** Sets the ISO 3166-1 region used to localize release information. */
    fun region(value: Region?) =
        apply {
            region = value
        }

    override fun toQueryParams(): QueryParams =
        QueryParams
            .create()
            .add("language", language?.value)
            .add("page", page)
            .add("region", region?.value)

    companion object {
        /** Creates an empty query that relies on TMDB defaults. */
        @JvmStatic
        fun create(): MovieUpcomingQuery = MovieUpcomingQuery()
    }
}
